File: code/linkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        //Floyd’s Cycle Finding Algorithm
        if(head == nullptr) return false;
        
        ListNode *slow = head;
        ListNode *fast = head;
        
        // checking if fast->next is also nullptr because if fast->next == nullptr, 
        // then fast->next->next causes segfault
        while(fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;
            
            if(fast == slow) return true;
        }
        
        // if fast or fast->next reaches nullptr, then it means no cycle.
        return false;
    }
};

File: code/linkedListCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        // using set store put ListNode
        Set<ListNode> listSet = new HashSet<ListNode>();
        int setSize = 0;
        
        if(head == null) return false;
        // since set cannot have duplicates, if the size stays the same after adding,
        // it means the node contains a cycle.
        while(head != null) {
            listSet.add(head);
            setSize++;
            if(listSet.size() != setSize) return true;
            head = head.next;
        }
        
        return false;
    }
}